File: weave/connector/connector.py
"""Local-filesystem connector for Claude Code session JSONL.

A dumb I/O boundary: translate a Claude Code session id <-> a file path on a
unix ~/.claude, and move bytes in and out of those paths. No JSON parsing, no
field rewriting, no session-id generation -- those live in the weave layer.

Storage base honors $CLAUDE_CONFIG_DIR if set, else ~/.claude. Stdlib only.
"""

import os
import re
import tempfile
from pathlib import Path

_ENCODE_RE = re.compile(r"[^A-Za-z0-9]")


class SessionNotFound(ValueError):
    """A read target (session id or path) does not exist."""


class AmbiguousSession(ValueError):
    """A session id matches files in more than one project directory."""


# --- path mechanics (pure, no I/O) -------------------------------------------
def projects_root():
    """`$CLAUDE_CONFIG_DIR/projects` if the env var is set, else
    `~/.claude/projects` (expanded). Read at call time."""
    base = os.environ.get("CLAUDE_CONFIG_DIR")
    root = Path(base) if base else Path("~/.claude")
    return root.expanduser() / "projects"


def encode_cwd(cwd):
    """The Claude Code project-dir encoding: every non-alphanumeric character
    becomes '-' (e.g. '/Users/bob/myapp' -> '-Users-bob-myapp')."""
    return _ENCODE_RE.sub("-", cwd)


def session_path(cwd, session_id):
    """Where a session for `cwd` with id `session_id` lives on this machine."""
    return projects_root() / encode_cwd(cwd) / f"{session_id}.jsonl"


# --- resolution / enumeration ------------------------------------------------
def resolve(session_id):
    """The path of the session with this id, or None. Raises AmbiguousSession
    if the id exists in more than one project dir (e.g. after a copy)."""
    matches = sorted(projects_root().glob(f"*/{session_id}.jsonl"))
    if not matches:
        return None
    if len(matches) > 1:
        dirs = ", ".join(m.parent.name for m in matches)
        raise AmbiguousSession(
            f"session id {session_id!r} found in {len(matches)} project "
            f"dirs: {dirs}")
    return matches[0]


def list_sessions():
    """(session_id, path) for every session file under the root, sorted by
    path. session_id is the filename stem."""
    return sorted(
        ((p.stem, p) for p in projects_root().glob("*/*.jsonl")),
        key=lambda pair: str(pair[1]))


def latest_session(cwd):
    """Session id of the most recently modified session for `cwd` (by mtime).

    Scoped to the single project directory that encodes `cwd`. Raises
    SessionNotFound when that project has no sessions yet."""
    proj = projects_root() / encode_cwd(cwd)
    files = list(proj.glob("*.jsonl")) if proj.is_dir() else []
    if not files:
        raise SessionNotFound(f"no local sessions for {cwd!r}")
    return max(files, key=lambda p: p.stat().st_mtime).stem


# --- bytes in ----------------------------------------------------------------
def read_text(session):
    """Read a session's JSONL as text. `session` is a session id OR a path:
    if it contains os.sep or ends with '.jsonl' it is treated as a path,
    otherwise as a session id to resolve. Raises SessionNotFound if absent."""
    if os.sep in session or session.endswith(".jsonl"):
        path = Path(session)
    else:
        path = resolve(session)
        if path is None:
            raise SessionNotFound(f"no session with id {session!r}")
    if not path.is_file():
        raise SessionNotFound(f"no session file at {str(path)!r}")
    return path.read_text(encoding="utf-8")


# --- bytes out ---------------------------------------------------------------
def write_text(path, text):
    """Atomically write `text` to `path`, creating parent dirs. Overwrites
    unconditionally and writes the string exactly as given. Returns the path.

    Atomic: a temp file in the same directory is written then os.replace()d
    into place, so an interrupted write never leaves a half-written file."""
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp = tempfile.mkstemp(
        dir=path.parent, prefix=path.name + ".", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8", newline="") as f:
            f.write(text)
        os.replace(tmp, path)
    except BaseException:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise
    return path

File: internal/connector/connector.go
// Package connector is a dumb I/O boundary for Claude Code session JSONL files.
package connector

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var encodeRE = regexp.MustCompile(`[^A-Za-z0-9]`)

// ErrSessionNotFound means a read target (session id or path) does not exist.
var ErrSessionNotFound = errors.New("session not found")

// ErrAmbiguousSession means a session id matches files in more than one project dir.
var ErrAmbiguousSession = errors.New("ambiguous session")

// ProjectsRoot returns $CLAUDE_CONFIG_DIR/projects if set, else ~/.claude/projects.
func ProjectsRoot() string {
	base := os.Getenv("CLAUDE_CONFIG_DIR")
	if base == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			base = filepath.Join("~", ".claude")
		} else {
			base = filepath.Join(home, ".claude")
		}
	}
	return filepath.Join(base, "projects")
}

// EncodeCWD is the Claude Code project-dir encoding: non-alphanumeric runes become '-'.
func EncodeCWD(cwd string) string {
	return encodeRE.ReplaceAllString(cwd, "-")
}

// SessionPath is where a session for cwd with sessionID lives on this machine.
func SessionPath(cwd, sessionID string) string {
	return filepath.Join(ProjectsRoot(), EncodeCWD(cwd), sessionID+".jsonl")
}

// Resolve returns the path of the session with this id, or "" when absent.
// It returns ErrAmbiguousSession when the id exists in more than one project dir.
func Resolve(sessionID string) (string, error) {
	root := ProjectsRoot()
	pattern := filepath.Join(root, "*", sessionID+".jsonl")
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return "", err
	}
	sort.Strings(matches)
	if len(matches) == 0 {
		return "", nil
	}
	if len(matches) > 1 {
		names := make([]string, len(matches))
		for i, m := range matches {
			names[i] = filepath.Base(filepath.Dir(m))
		}
		return "", fmt.Errorf("%w: session id %q found in %d project dirs: %s",
			ErrAmbiguousSession, sessionID, len(matches), strings.Join(names, ", "))
	}
	return matches[0], nil
}

// Session is one local session file (id and absolute or relative path).
type Session struct {
	ID   string
	Path string
}

// ListSessions returns every session file under the root, sorted by path.
func ListSessions() ([]Session, error) {
	root := ProjectsRoot()
	pattern := filepath.Join(root, "*", "*.jsonl")
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return nil, err
	}
	sort.Strings(matches)
	out := make([]Session, len(matches))
	for i, p := range matches {
		out[i] = Session{ID: strings.TrimSuffix(filepath.Base(p), ".jsonl"), Path: p}
	}
	return out, nil
}

// ListForCWD returns sessions for one cwd project dir, sorted by path.
func ListForCWD(cwd string) ([]Session, error) {
	proj := filepath.Join(ProjectsRoot(), EncodeCWD(cwd))
	pattern := filepath.Join(proj, "*.jsonl")
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return nil, err
	}
	sort.Strings(matches)
	out := make([]Session, len(matches))
	for i, p := range matches {
		out[i] = Session{ID: strings.TrimSuffix(filepath.Base(p), ".jsonl"), Path: p}
	}
	return out, nil
}

// ReadText reads a session's JSONL. session is a session id or a path: if it
// contains a path separator or ends with ".jsonl" it is treated as a path.
func ReadText(session string) (string, error) {
	var path string
	if strings.Contains(session, string(os.PathSeparator)) || strings.HasSuffix(session, ".jsonl") {
		path = session
	} else {
		resolved, err := Resolve(session)
		if err != nil {
			return "", err
		}
		if resolved == "" {
			return "", fmt.Errorf("%w: no session with id %q", ErrSessionNotFound, session)
		}
		path = resolved
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return "", fmt.Errorf("%w: no session file at %q", ErrSessionNotFound, path)
		}
		return "", err
	}
	return string(data), nil
}

// WriteText atomically writes text to path, creating parent dirs. Overwrites
// unconditionally and writes the string exactly as given. Returns path.
func WriteText(path, text string) (string, error) {
	path = filepath.Clean(path)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return "", err
	}
	tmp, err := os.CreateTemp(filepath.Dir(path), filepath.Base(path)+".*.tmp")
	if err != nil {
		return "", err
	}
	tmpPath := tmp.Name()
	ok := false
	defer func() {
		if !ok {
			_ = os.Remove(tmpPath)
		}
	}()
	if _, err := tmp.WriteString(text); err != nil {
		_ = tmp.Close()
		return "", err
	}
	if err := tmp.Close(); err != nil {
		return "", err
	}
	if err := os.Rename(tmpPath, path); err != nil {
		return "", err
	}
	ok = true
	return path, nil
}
